[One Workflow] JSON Schema based inputs

h88 opened the pull request:
## Summary

Adds support for JSON Schema-based workflow inputs alongside the existing array format, with comprehensive validation, autocomplete, and backward compatibility. This implementation is based on the [JSON Schema specification](https://json-schema.org/)

## Features

### JSON Schema Inputs
- Support for JSON Schema format (`properties`, `required`, `additionalProperties`, `definitions`, `$defs`) based on the [JSON Schema specification](https://json-schema.org/)
- Backward compatibility with legacy array format inputs
- Automatic normalization between formats

### Validation
- **Monaco Editor**: Live validation of `default` values against JSON Schema constraints (type, format, pattern, minLength, etc.)
- **Pre-run Modal**: Input validation before workflow execution
- **$ref Support**: Validation for referenced schemas in `definitions` and `$defs`
- Accurate line/column error positioning in Monaco editor

### Autocomplete
- Type suggestions (`string`, `number`, `boolean`, `object`, `array`)
- Format suggestions (`email`, `date-time`, `uri`, etc.)
- Enum value suggestions from JSON Schema `enum` definitions
- Property key suggestions for nested objects

### Type Safety
- Recursive Zod types for nested JSON Schema objects
- Proper type inference for workflow context inputs
- Type-safe conversion utilities

## Technical Details

- Added `validateJsonSchemaDefaults()` for Monaco editor validation
- Added `getJsonSchemaSuggestions()` for autocomplete support
- Enhanced `normalizeInputsToJsonSchema()` with `$ref` resolution
- Updated `WorkflowContextSchema` to support nested objects
- Refactored code to reduce complexity

## Testing

- Unit tests for validation logic
- Unit tests for autocomplete suggestions
- Integration tests for backward compatibility
- Manual testing in Monaco editor and execution modal

## Backward Compatibility

 - Legacy array format inputs continue to work
 - Existing workflows are not affected
 - Format conversion only occurs when explicitly updating inputs

## Related Issues

Fixes [elastic/security-team#14734](https://github.com/elastic/security-team/issues/14734)
Fixes https://github.com/elastic/security-team/issues/14675




https://github.com/user-attachments/assets/193e4bbd-bf37-4870-90d3-48831381c747


<details>
<summary>Design</summary>

## Component Details

### 1. **Normalization Layer** (Core)
**Location**: `@kbn/workflows/spec/lib/input_conversion.ts`

**Key Functions**:
- `normalizeInputsToJsonSchema()`: Entry point - handles both legacy array and JSON Schema formats
- `convertLegacyInputsToJsonSchema()`: Converts legacy array format to JSON Schema
- `resolveRef()`: Resolves `$ref` references to definitions/$defs
- `applyInputDefaults()`: Recursively applies default values from JSON Schema
- `applyDefaultFromSchema()`: Handles nested objects and $ref for defaults

**Purpose**: Single source of truth for input format conversion and normalization

### 2. **Schema Definition Layer**
**Location**: `@kbn/workflows/spec/schema.ts`

**Key Components**:
- `WorkflowInputsJsonSchema`: Zod schema for JSON Schema format
- `WorkflowInputArraySchema`: Legacy array format (for backward compatibility)
- `WorkflowSchema`: Main schema with `.pipe()` transformation
  - Accepts both formats via `z.union()`
  - Transforms legacy array → JSON Schema via `.transform()`
  - Outputs only JSON Schema format via `.pipe()`

**Purpose**: Type-safe schema definition with automatic transformation

### 3. **Monaco Editor Validation**
**Location**: `workflows_management/public/features/validate_workflow_yaml/lib/validate_json_schema_defaults.ts`

**Flow**:
1. YAML document parsed with `yaml` library
2. `visit()` function traverses YAML AST
3. Finds all `default:` scalar nodes
4. Extracts property path from ancestors
5. Normalizes inputs to JSON Schema
6. Resolves $ref if needed
7. Converts JSON Schema to Zod
8. Validates default value against Zod schema
9. Reports errors with accurate line/column (via `model.getPositionAt()`)

**Purpose**: Real-time validation in Monaco editor

### 4. **Pre-Run Modal Validation**
**Location**: `workflows_management/public/features/run_workflow/ui/workflow_execute_manual_form.tsx`

**Flow**:
1. `makeWorkflowInputsValidator()` creates Zod validator
2. Normalizes inputs to JSON Schema
3. Converts each property to Zod schema (with $ref resolution)
4. Validates user input on change and initial load
5. Applies defaults using `applyInputDefaults()`

**Purpose**: Form validation before workflow execution

### 5. **Schema Generation for Monaco**
**Location**: `@kbn/workflows/spec/lib/get_json_schema_from_yaml_schema.ts`

**Flow**:
1. `generateYamlSchemaFromConnectors()` creates Zod schema
2. `getJsonSchemaFromYamlSchema()` converts Zod → JSON Schema
3. `fixInputsSchemaForMonaco()` ensures backward compatibility
   - Handles `allOf` structure from `.pipe()` transformation
   - Adds `anyOf` to support both array and object formats
4. Result: JSON Schema that Monaco can use for autocomplete/validation

**Purpose**: Generate Monaco-compatible JSON Schema with backward compatibility

### 6. **Runtime Context Schema**
**Location**: `workflows_management/public/features/workflow_context/lib/get_workflow_context_schema.ts`

**Flow**:
1. Normalizes inputs to JSON Schema
2. Converts each property to Zod schema
3. Builds `DynamicWorkflowContextSchema` with inputs
4. Used for variable validation in Liquid templates

**Purpose**: Runtime validation of workflow variables

### 7. **Execution Layer**
**Location**: `workflows_execution_engine/server/workflow_context_manager/build_workflow_context.ts`

**Flow**:
1. Normalizes inputs from workflow definition
2. Applies defaults using `applyInputDefaults()`
3. Recursively handles nested objects and $ref
4. Builds `WorkflowContext` with inputs + defaults
5. Used during workflow execution

**Purpose**: Apply defaults at execution time

### 8. **External Integration: OneChat**
**Location**: `x-pack/platform/plugins/shared/onechat/server/services/tools/tool_types/workflow/generate_schema.ts`

**Flow**:
1. Receives workflow definition
2. Normalizes inputs to JSON Schema
3. Converts JSON Schema to Zod (simple converter)
4. Generates tool schema for AI agents

**Purpose**: Enable workflows as tools in AI chat

## Data Flow Examples

### Example 1: Legacy Array Input
```yaml
inputs:
  - name: email
    type: string
    default: "user@example.com"
    required: true
```

**Flow**:
1. YAML parsed → `WorkflowSchema` accepts array
2. `.transform()` → `convertLegacyInputsToJsonSchema()`
3. Result:
```json
{
  "properties": {
    "email": {
      "type": "string",
      "default": "user@example.com"
    }
  },
  "required": ["email"]
}
```

### Example 2: JSON Schema with $ref
```yaml
inputs:
  properties:
    user:
      $ref: "#/definitions/UserSchema"
  definitions:
    UserSchema:
      type: object
      properties:
        email:
          type: string
          format: email
          default: "user@example.com"
```

**Flow**:
1. YAML parsed → Already JSON Schema format
2. `normalizeInputsToJsonSchema()` returns as-is
3. `resolveRef()` resolves `$ref` to `UserSchema`
4. Validation uses resolved schema
5. Defaults applied recursively

### Example 3: Nested Object Defaults
```yaml
inputs:
  properties:
    analyst:
      type: object
      properties:
        email:
          type: string
          format: email
          default: "analyst@example.com"
```

**Flow**:
1. Normalized to JSON Schema
2. `applyInputDefaults()` recursively applies defaults
3. `applyDefaultFromSchema()` handles nested structure
4. Result: `{ analyst: { email: "analyst@example.com" } }`

## Key Design Decisions

1. **Single Normalization Point**: `normalizeInputsToJsonSchema()` is the single source of truth
2. **Backward Compatibility**: Legacy array format automatically converted, never lost
3. **Type Safety**: Zod schemas ensure type safety throughout
4. **$ref Support**: Full support for JSON Schema references
5. **Recursive Defaults**: Nested objects and arrays fully supported
6. **Multiple Validation Points**: Editor, modal, and runtime validation
7. **Accurate Error Reporting**: Monaco uses `model.getPositionAt()` for precise line/column

## File Structure

```
kbn-workflows/spec/
├── schema.ts                          # Zod schema definitions
├── lib/
│   ├── input_conversion.ts           # Core normalization & defaults
│   ├── get_json_schema_from_yaml_schema.ts  # Monaco schema generation
│   └── generate_yaml_schema_from_connectors.ts  # Schema generation

workflows_management/
├── public/features/
│   ├── validate_workflow_yaml/
│   │   └── lib/validate_json_schema_defaults.ts  # Monaco validation
│   ├── run_workflow/
│   │   └── ui/workflow_execute_manual_form.tsx  # Modal validation
│   └── workflow_context/
│       └── lib/get_workflow_context_schema.ts  # Runtime validation
└── server/
    └── workflows_management_service.ts  # Preserve format on save

workflows_execution_engine/
└── server/workflow_context_manager/
    └── build_workflow_context.ts  # Execution-time defaults

x-pack/platform/plugins/shared/onechat/
└── server/services/tools/tool_types/workflow/
    └── generate_schema.ts  # External integration
```




</details>

h88 committed:
JSON Schema based 1Workflow inputs - init

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

h88 committed:
minor refactor - complexity - conversion & scalar methods are now simpler, delegating to helpers

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/365536)
* Commit: 47ca913de5c05956adda7f230520daaf29533ccf

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/365536#019abf5a-dba4-44fc-b224-1262b45d7f43)

### History
* :broken_heart: [Build #365299](https://buildkite.com/elastic/kibana-pull-request/builds/365299) failed e931e913bc5de2784d6e51e32a4d4eb66079e6f7

cc @h88

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365536","number":365536,"commit":"47ca913de5c05956adda7f230520daaf29533ccf"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365299","number":365299,"commit":"e931e913bc5de2784d6e51e32a4d4eb66079e6f7"}],"number":365536}
buildkite-pr-comment-->

h88 committed:
typecheck / WorkflowZodSchemaType schema type instead of inferred type

h88 committed:
ci / fix ts errors in test files

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

h88 committed:
ut fix

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/366341)
* Commit: 93d552e73fe9e41b919ebf792c34f2c840ea74b0

### Failed CI Steps
* [Jest Tests #17](https://buildkite.com/elastic/kibana-pull-request/builds/366341#019ac58b-2c51-49df-a3ad-a07987de2ec1)
* [Jest Tests #17](https://buildkite.com/elastic/kibana-pull-request/builds/366341#019ac59a-a9d1-41fc-a60f-abf3ead0d302)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/366341#019ac58b-2c51-49df-a3ad-a07987de2ec1) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/366341/jobs/019ac58b-2c51-49df-a3ad-a07987de2ec1/artifacts/019ac59a-8bc4-4059-8915-3829ab253e68) Jest Tests #<span></span>17 / Generated Kibana Connectors Nested Schema Issue Detection should fix Monaco JSON schema generation to restore YAML validation
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/366341#019ac59a-a9d1-41fc-a60f-abf3ead0d302) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/366341/jobs/019ac59a-a9d1-41fc-a60f-abf3ead0d302/artifacts/019ac5a8-9dbb-46ee-b3d1-30f40e2e844c) Jest Tests #<span></span>17 / Generated Kibana Connectors Nested Schema Issue Detection should fix Monaco JSON schema generation to restore YAML validation

### History
* :broken_heart: [Build #366266](https://buildkite.com/elastic/kibana-pull-request/builds/366266) failed 4c676b60d0cfa3f968e371104523a0749602bb47
* :broken_heart: [Build #366161](https://buildkite.com/elastic/kibana-pull-request/builds/366161) failed f3fb067bb6d9d0672a2b32dac4a8b662cb5a44d7
* :broken_heart: [Build #365536](https://buildkite.com/elastic/kibana-pull-request/builds/365536) failed 47ca913de5c05956adda7f230520daaf29533ccf
* :broken_heart: [Build #365299](https://buildkite.com/elastic/kibana-pull-request/builds/365299) failed e931e913bc5de2784d6e51e32a4d4eb66079e6f7

cc @h88

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366341","number":366341,"commit":"93d552e73fe9e41b919ebf792c34f2c840ea74b0"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366266","number":366266,"commit":"4c676b60d0cfa3f968e371104523a0749602bb47"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366161","number":366161,"commit":"f3fb067bb6d9d0672a2b32dac4a8b662cb5a44d7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365536","number":365536,"commit":"47ca913de5c05956adda7f230520daaf29533ccf"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365299","number":365299,"commit":"e931e913bc5de2784d6e51e32a4d4eb66079e6f7"}],"number":366341}
buildkite-pr-comment-->

h88 committed:
ut fix

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

# Conflicts:
#	src/platform/plugins/shared/workflows_management/public/features/run_workflow/ui/workflow_execute_manual_form.tsx

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/366910)
* Commit: 2699e9ce84bdb09feb518c6a889301c79a1f2d9e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/d0595b1fc31ba8cc0a3a4ea880264f770306a17f) | [after](https://github.com/elastic/kibana/commit/2699e9ce84bdb09feb518c6a889301c79a1f2d9e) | diff |
| --- | --- | --- | --- |
| `workflowsManagement` | 809 | 927 | +118 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/d0595b1fc31ba8cc0a3a4ea880264f770306a17f) | [after](https://github.com/elastic/kibana/commit/2699e9ce84bdb09feb518c6a889301c79a1f2d9e) | diff |
| --- | --- | --- | --- |
| `@kbn/workflows` | 335 | 336 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/d0595b1fc31ba8cc0a3a4ea880264f770306a17f) | [after](https://github.com/elastic/kibana/commit/2699e9ce84bdb09feb518c6a889301c79a1f2d9e) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 1.0MB | 1.0MB | +38.0B |
| `workflowsManagement` | 2.1MB | 2.2MB | +150.9KB |
| total |  |  | +150.9KB |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/d0595b1fc31ba8cc0a3a4ea880264f770306a17f) | [after](https://github.com/elastic/kibana/commit/2699e9ce84bdb09feb518c6a889301c79a1f2d9e) | diff |
| --- | --- | --- | --- |
| `@kbn/workflows` | 365 | 366 | +1 |

#### ESLint disabled in files
| id | [before](https://github.com/elastic/kibana/commit/d0595b1fc31ba8cc0a3a4ea880264f770306a17f) | [after](https://github.com/elastic/kibana/commit/2699e9ce84bdb09feb518c6a889301c79a1f2d9e) | diff |
| --- | --- | --- | --- |
| `workflowsManagement` | 38 | 36 | -2 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/d0595b1fc31ba8cc0a3a4ea880264f770306a17f) | [after](https://github.com/elastic/kibana/commit/2699e9ce84bdb09feb518c6a889301c79a1f2d9e) | diff |
| --- | --- | --- | --- |
| `@kbn/workflows` | 1178 | 1177 | -1 |
| `workflowsManagement` | 85 | 94 | +9 |
| total |  |  | +8 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/d0595b1fc31ba8cc0a3a4ea880264f770306a17f) | [after](https://github.com/elastic/kibana/commit/2699e9ce84bdb09feb518c6a889301c79a1f2d9e) | diff |
| --- | --- | --- | --- |
| `@kbn/workflows` | 1178 | 1177 | -1 |
| `workflowsManagement` | 123 | 130 | +7 |
| total |  |  | +6 |


</details>


### History
* :yellow_heart: [Build #366432](https://buildkite.com/elastic/kibana-pull-request/builds/366432) was flaky 72481499e91a62b325f1006af8d42dede200c75d
* :broken_heart: [Build #366341](https://buildkite.com/elastic/kibana-pull-request/builds/366341) failed 93d552e73fe9e41b919ebf792c34f2c840ea74b0
* :broken_heart: [Build #366266](https://buildkite.com/elastic/kibana-pull-request/builds/366266) failed 4c676b60d0cfa3f968e371104523a0749602bb47
* :broken_heart: [Build #366161](https://buildkite.com/elastic/kibana-pull-request/builds/366161) failed f3fb067bb6d9d0672a2b32dac4a8b662cb5a44d7
* :broken_heart: [Build #365536](https://buildkite.com/elastic/kibana-pull-request/builds/365536) failed 47ca913de5c05956adda7f230520daaf29533ccf
* :broken_heart: [Build #365299](https://buildkite.com/elastic/kibana-pull-request/builds/365299) failed e931e913bc5de2784d6e51e32a4d4eb66079e6f7

cc @h88

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366910","number":366910,"commit":"2699e9ce84bdb09feb518c6a889301c79a1f2d9e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366432","number":366432,"commit":"72481499e91a62b325f1006af8d42dede200c75d"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366341","number":366341,"commit":"93d552e73fe9e41b919ebf792c34f2c840ea74b0"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366266","number":366266,"commit":"4c676b60d0cfa3f968e371104523a0749602bb47"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366161","number":366161,"commit":"f3fb067bb6d9d0672a2b32dac4a8b662cb5a44d7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365536","number":365536,"commit":"47ca913de5c05956adda7f230520daaf29533ccf"}],"number":366910}
buildkite-pr-comment-->

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

h88 committed:
?? merge issues

h88 committed:
irrelevant

h88 committed:
?? merge gh vs local git

h88 committed:
restore back removed backward compatibility tests for JSON schema generation

Updated tests to handle:
- allOf schema structure from .pipe() transformations
- anyOf union structures in connector 'with' properties
- Schema navigation for both direct properties and nested allOf items

h88 committed:
ref rep JsonSchemaObject type

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

# Conflicts:
#	src/platform/plugins/shared/workflows_management/public/pages/workflow_detail/ui/workflow_detail_test_modal.test.tsx

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/367469)
* Commit: 8dd9010caea44f69dd7a47a0e0442817523e02a6

### Failed CI Steps
* [Agent Builder API Smoke Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/367469#019ade1e-44ae-4de6-b8af-e1de2c788d72)
* [Agent Builder API Smoke Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/367469#019ade3d-0c23-43b5-88f3-404061ab1b81)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/367469#019ade1e-44ae-4de6-b8af-e1de2c788d72) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/367469/jobs/019ade1e-44ae-4de6-b8af-e1de2c788d72/artifacts/019ade3c-fa8e-43b2-8ae3-49482a1e5085) Agent Builder API Smoke Tests #<span></span>1 / Agent Builder - LLM Smoke tests Connector "gemini-2-flash" Converse API sync can execute a tool
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/367469#019ade3d-0c23-43b5-88f3-404061ab1b81) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/367469/jobs/019ade3d-0c23-43b5-88f3-404061ab1b81/artifacts/019ade45-f4cc-4ea2-a648-59f3366cd481) Agent Builder API Smoke Tests #<span></span>1 / Agent Builder - LLM Smoke tests Connector "gemini-2-flash" Converse API sync can execute a tool

### History
* :green_heart: [Build #367072](https://buildkite.com/elastic/kibana-pull-request/builds/367072) succeeded 7f1d0b59447919ff7d8a848f44a416f2e3b64e52
* :yellow_heart: [Build #366932](https://buildkite.com/elastic/kibana-pull-request/builds/366932) was flaky 014c3c52e9586178bbbcae1bde9d3b02a39f467d
* :green_heart: [Build #366910](https://buildkite.com/elastic/kibana-pull-request/builds/366910) succeeded 2699e9ce84bdb09feb518c6a889301c79a1f2d9e
* :yellow_heart: [Build #366432](https://buildkite.com/elastic/kibana-pull-request/builds/366432) was flaky 72481499e91a62b325f1006af8d42dede200c75d
* :broken_heart: [Build #366341](https://buildkite.com/elastic/kibana-pull-request/builds/366341) failed 93d552e73fe9e41b919ebf792c34f2c840ea74b0
* :broken_heart: [Build #366266](https://buildkite.com/elastic/kibana-pull-request/builds/366266) failed 4c676b60d0cfa3f968e371104523a0749602bb47

cc @h88

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367469","number":367469,"commit":"8dd9010caea44f69dd7a47a0e0442817523e02a6"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367072","number":367072,"commit":"7f1d0b59447919ff7d8a848f44a416f2e3b64e52"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366932","number":366932,"commit":"014c3c52e9586178bbbcae1bde9d3b02a39f467d"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366910","number":366910,"commit":"2699e9ce84bdb09feb518c6a889301c79a1f2d9e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366432","number":366432,"commit":"72481499e91a62b325f1006af8d42dede200c75d"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366341","number":366341,"commit":"93d552e73fe9e41b919ebf792c34f2c840ea74b0"}],"number":367469}
buildkite-pr-comment-->

h88 committed:
defensive null checks for partially parsed YAML

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/368018)
* Commit: 9bb2b8c5cf309165a443096fd04057616061ab9b

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/368018#019ae122-b821-4952-88d0-1887d267d0fd)

### History
* :broken_heart: [Build #367469](https://buildkite.com/elastic/kibana-pull-request/builds/367469) failed 8dd9010caea44f69dd7a47a0e0442817523e02a6
* :green_heart: [Build #367072](https://buildkite.com/elastic/kibana-pull-request/builds/367072) succeeded 7f1d0b59447919ff7d8a848f44a416f2e3b64e52
* :yellow_heart: [Build #366932](https://buildkite.com/elastic/kibana-pull-request/builds/366932) was flaky 014c3c52e9586178bbbcae1bde9d3b02a39f467d
* :green_heart: [Build #366910](https://buildkite.com/elastic/kibana-pull-request/builds/366910) succeeded 2699e9ce84bdb09feb518c6a889301c79a1f2d9e
* :yellow_heart: [Build #366432](https://buildkite.com/elastic/kibana-pull-request/builds/366432) was flaky 72481499e91a62b325f1006af8d42dede200c75d

cc @h88

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368018","number":368018,"commit":"9bb2b8c5cf309165a443096fd04057616061ab9b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367469","number":367469,"commit":"8dd9010caea44f69dd7a47a0e0442817523e02a6"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367072","number":367072,"commit":"7f1d0b59447919ff7d8a848f44a416f2e3b64e52"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366932","number":366932,"commit":"014c3c52e9586178bbbcae1bde9d3b02a39f467d"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366910","number":366910,"commit":"2699e9ce84bdb09feb518c6a889301c79a1f2d9e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366432","number":366432,"commit":"72481499e91a62b325f1006af8d42dede200c75d"}],"number":368018}
buildkite-pr-comment-->

h88 committed:
nested objects defense

h88 commented on src/platform/packages/shared/kbn-workflows/spec/schema.ts:
```diff
@@ -416,6 +418,53 @@ export const WorkflowInputSchema = z.union([
 
+// New JSON Schema-based inputs structure
+// This represents a JSON Schema object with properties, required, additionalProperties, and definitions
+export const WorkflowInputsJsonSchema = z
```
TBC

h88 committed:
tsc --fix

h88 committed:
tsc --fix

jbudz commented on packages/kbn-optimizer/src/worker/webpack.config.ts:
```diff
@@ -96,6 +96,22 @@ export function getWebpackConfig(
       }),
+      // Fix for @n8n/json-schema-to-zod ESM imports without extensions
+      // This package uses ESM imports without file extensions, which webpack 5 requires
+      new webpack.NormalModuleReplacementPlugin(/^\.\.?\/.*$/, (resource: any) => {
```
Did we consider creating a rule that unsets fullySpecified for this package?  It seems like this would be a more idiomatic approach:

https://webpack.js.org/configuration/resolve/#resolvefullyspecified
```
{
  test: ...,
  include: ..., 
  resolve: {
    fullySpecified: false
  },
},
```

Kiryous commented on src/platform/plugins/shared/workflows_management/public/widgets/workflow_yaml_editor/lib/autocomplete/suggestions/get_suggestions.ts:
```diff
@@ -136,18 +137,22 @@ export function getSuggestions(
+  //       enum: |<- (suggest enum values from schema)
+  // This should be checked BEFORE other type completions to avoid conflicts
+  // but AFTER variable/connector completions which are more specific
+  const jsonSchemaSuggestions = getJsonSchemaSuggestions(autocompleteContext);
```
I agree. @h88 did you try to rely on monaco-yaml completions?

Kiryous replied:
For this to work we would need a detailed zod schema representing all json schema types in `src/platform/packages/shared/kbn-workflows/spec/schema.ts` instead of z.any():

```
export const WorkflowInputsJsonSchema = z
  .object({
    properties: z.record(z.string(), z.any()).optional(),
    required: z.array(z.string()).optional(),
    additionalProperties: z.union([z.boolean(), z.any()]).optional(),
    definitions: z.record(z.string(), z.any()).optional(),
    $defs: z.record(z.string(), z.any()).optional(),
  })
```

but it might inflate the workflow json schema too much, we need to see.

skynetigor replied:
so getJsonSchemaSuggestions is called for steps?

h88 committed:
Refactor replace NormalModuleReplacementPlugin with webpack module rule using resolve.fullySpecified

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/369013)
* Commit: b1761fc00b2aa99f32375202d1d6b321093622a1

### Failed CI Steps
* [FTR Configs #35](https://buildkite.com/elastic/kibana-pull-request/builds/369013#019ae9b1-198a-479d-8057-037f03b8a83e)
* [FTR Configs #35](https://buildkite.com/elastic/kibana-pull-request/builds/369013#019ae9d9-d4d4-48d1-a7f5-a5f839e725c0)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/369013#019ae9b1-198a-479d-8057-037f03b8a83e) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/369013/jobs/019ae9b1-198a-479d-8057-037f03b8a83e/artifacts/019ae9d9-a972-4d2d-af05-533fc15e4aa3) FTR Configs #<span></span>35 / Maps endpoints apis search ES|QL should return getValues response in expected shape
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/369013#019ae9d9-d4d4-48d1-a7f5-a5f839e725c0) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/369013/jobs/019ae9d9-d4d4-48d1-a7f5-a5f839e725c0/artifacts/019ae9e0-cb28-48a2-9333-be75b6e5d96f) FTR Configs #<span></span>35 / Maps endpoints apis search ES|QL should return getValues response in expected shape

### History
* :broken_heart: [Build #368493](https://buildkite.com/elastic/kibana-pull-request/builds/368493) failed c5f9c1e9263ab0d6050ec384fe8d5995d17df12a
* :broken_heart: [Build #368237](https://buildkite.com/elastic/kibana-pull-request/builds/368237) failed 4b6f2b20dd2fd5fde33ef6bcd4770aaa39fdf078
* :broken_heart: [Build #368022](https://buildkite.com/elastic/kibana-pull-request/builds/368022) failed 66fc2b43b5f4eee6e4e20135851c3b914d6a1f99
* :broken_heart: [Build #368018](https://buildkite.com/elastic/kibana-pull-request/builds/368018) failed 9bb2b8c5cf309165a443096fd04057616061ab9b
* :broken_heart: [Build #367469](https://buildkite.com/elastic/kibana-pull-request/builds/367469) failed 8dd9010caea44f69dd7a47a0e0442817523e02a6

cc @h88

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369013","number":369013,"commit":"b1761fc00b2aa99f32375202d1d6b321093622a1"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368493","number":368493,"commit":"c5f9c1e9263ab0d6050ec384fe8d5995d17df12a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368237","number":368237,"commit":"4b6f2b20dd2fd5fde33ef6bcd4770aaa39fdf078"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368022","number":368022,"commit":"66fc2b43b5f4eee6e4e20135851c3b914d6a1f99"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368018","number":368018,"commit":"9bb2b8c5cf309165a443096fd04057616061ab9b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367469","number":367469,"commit":"8dd9010caea44f69dd7a47a0e0442817523e02a6"}],"number":369013}
buildkite-pr-comment-->

h88 committed:
Add null check in convertLegacyInputsToJsonSchema

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs "tough times never last, only tough people last"

# Conflicts:
#	src/platform/packages/shared/kbn-workflows/common/generated/kibana_connectors.test.ts
#	src/platform/packages/shared/kbn-workflows/spec/lib/generate_yaml_schema_from_connectors.ts
#	src/platform/packages/shared/kbn-workflows/spec/lib/get_json_schema_from_yaml_schema.test.ts
#	src/platform/packages/shared/kbn-workflows/spec/lib/get_json_schema_from_yaml_schema.ts
#	src/platform/packages/shared/kbn-workflows/spec/schema.ts
#	src/platform/plugins/shared/workflows_management/common/lib/yaml/parse_workflow_yaml_to_json.test.ts
#	src/platform/plugins/shared/workflows_management/common/lib/zod/zod_utils.test.ts
#	src/platform/plugins/shared/workflows_management/common/lib/zod/zod_utils.ts
#	src/platform/plugins/shared/workflows_management/common/schema.ts
#	src/platform/plugins/shared/workflows_management/public/features/run_workflow/ui/workflow_execute_manual_form.tsx
#	src/platform/plugins/shared/workflows_management/public/features/workflow_context/lib/get_workflow_context_schema.test.ts
#	src/platform/plugins/shared/workflows_management/public/features/workflow_context/lib/get_workflow_context_schema.ts
#	src/platform/plugins/shared/workflows_management/server/workflows_management/workflows_management_service.ts
#	x-pack/platform/plugins/shared/onechat/server/services/tools/tool_types/workflow/generate_schema.ts

h88 committed:
refactor / rename WorkflowInputsJsonSchema to JsonModelSchema

h88 committed:
lint --fix

h88 committed:
type_check --fix

h88 committed:
minor ac fix

h88 committed:
ut --fix

h88 committed:
fix: adapt JSON schema tests for transform-based workflow schemas

Refactored WorkflowSchema to use WorkflowSchemaBase for extension,
enabling generateYamlSchemaFromConnectors to extend the base schema
before applying transforms for inputs backward compatibility.

Updated JSON schema tests to handle transform schemas which may
generate different structures (e.g., $ref at root instead of inline
properties) when using reused: 'ref'. Tests now verify schema validity
via AJV compilation rather than assuming a specific structure, ensuring
compatibility with transform schemas while maintaining validation
correctness.

Changes:
- Introduced WorkflowSchemaBase (without transform) for extension
- Updated generateYamlSchemaFromConnectors to use WorkflowSchemaBase
- Made tests flexible to handle $ref at root or inline properties
- Removed $schema property before AJV compilation (AJV 2020 compatibility)
- Updated test assertions to be conditional on schema structure

This change is necessary because transforms are required to normalize
legacy array inputs to JSON Schema format, and z.toJSONSchema on
transform schemas may produce different structures than plain z.object()
schemas.

h88 committed:
fix: validate default values for legacy array format inputs

The validateJsonSchemaDefaults function was only handling the new JSON
Schema format (inputs.properties.*) and silently skipping validation for
legacy array format inputs (inputs[0].default). This fix adds support for
legacy array format by:

1. Detecting legacy array paths in extractPropertyPath (path[1] is a number)
2. Mapping array indices to property names using the original inputs array
3. Updating path length check to allow both formats (length 3 for legacy, 4 for new)

Since normalizeInputsToJsonSchema already converts legacy inputs to JSON
Schema format, the property schemas are already available - we just need
to map the YAML path correctly.

Added test cases to verify the fix works for both valid and invalid
default values in legacy array format.

h88 committed:
refactor: replace ZodTypeAny with ZodType per Zod v4 doc

ZodTypeAny has been eliminated in Zod v4. Updated return types and function signatures to use ZodType instead, which is the correct base type for all Zod schemas. No functional changes.

h88 committed:
bad merge residue debug > info

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

h88 committed:
Proper AC & inline validations.

Unwrap ZodPipe schemas (returned by generateYamlSchemaFromConnectors) before calling
z.toJSONSchema to generate proper JSON Schema with properties at root level. This
fixes missing autocompletion for workflow keys and missing validation errors for
invalid values (e.g., enabled: 23).

- Unwrap ZodPipe by accessing _def.in to get input schema
- Allow Monaco YAML validation/autocompletion when workflowDefinition is null
- Add regression test for ZodPipe unwrapping

h88 committed:
check_types / local

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/370624)
* Commit: a40e23766e12f69c1ec69a0eee585d76132a9e6e

### Failed CI Steps
* [Jest Tests #12](https://buildkite.com/elastic/kibana-pull-request/builds/370624#019b03ae-b741-4921-8446-a919b4d003da)
* [Jest Tests #12](https://buildkite.com/elastic/kibana-pull-request/builds/370624#019b03cd-d090-4b54-9bef-88f6eccb71a1)
* [FTR Configs #120](https://buildkite.com/elastic/kibana-pull-request/builds/370624#019b03ae-b7df-4d5a-b62e-b2c3e15c0753)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/370624#019b03ae-b741-4921-8446-a919b4d003da) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/370624/jobs/019b03ae-b741-4921-8446-a919b4d003da/artifacts/019b03cd-9fcb-425d-a42c-215a4eb9d759) Jest Tests #<span></span>12 / Monaco Schema Generation - Inputs Field should validate a workflow WITHOUT inputs (inputs field missing) - regression test
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/370624#019b03cd-d090-4b54-9bef-88f6eccb71a1) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/370624/jobs/019b03cd-d090-4b54-9bef-88f6eccb71a1/artifacts/019b03e8-c5e8-46da-9c7a-9526673a44f9) Jest Tests #<span></span>12 / Monaco Schema Generation - Inputs Field should validate a workflow WITHOUT inputs (inputs field missing) - regression test
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/370624#019b03ae-b741-4921-8446-a919b4d003da) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/370624/jobs/019b03ae-b741-4921-8446-a919b4d003da/artifacts/019b03cd-9fca-47e7-aec8-146b5be30b98) Jest Tests #<span></span>12 / Validate example_security_workflow.yaml against Monaco Schema should validate the security workflow YAML against the generated Monaco schema
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/370624#019b03cd-d090-4b54-9bef-88f6eccb71a1) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/370624/jobs/019b03cd-d090-4b54-9bef-88f6eccb71a1/artifacts/019b03e8-c5e7-45eb-9107-210e649b8af6) Jest Tests #<span></span>12 / Validate example_security_workflow.yaml against Monaco Schema should validate the security workflow YAML against the generated Monaco schema

### History
* :broken_heart: [Build #370587](https://buildkite.com/elastic/kibana-pull-request/builds/370587) failed f5e4820dee47a1b5bfb8282cb37fcbd7ebefc292
* :yellow_heart: [Build #370432](https://buildkite.com/elastic/kibana-pull-request/builds/370432) was flaky a868dbdde8d02690210420f315457c35e3aa3331
* :broken_heart: [Build #370184](https://buildkite.com/elastic/kibana-pull-request/builds/370184) failed 08a46335cd3a4e04c0c2203349b271dbee9ba9bd
* :broken_heart: [Build #369889](https://buildkite.com/elastic/kibana-pull-request/builds/369889) failed d7eb4e824870baf5010d2d432dec3fd05bb26a05
* :broken_heart: [Build #369876](https://buildkite.com/elastic/kibana-pull-request/builds/369876) failed 02ff815e179f596431a507ecc80b22fa1c648b32
* :broken_heart: [Build #369109](https://buildkite.com/elastic/kibana-pull-request/builds/369109) failed 0808d9845e7e278650dbeee5d174485cf550f8c4

cc @h88

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/370624","number":370624,"commit":"a40e23766e12f69c1ec69a0eee585d76132a9e6e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/370587","number":370587,"commit":"f5e4820dee47a1b5bfb8282cb37fcbd7ebefc292"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/370432","number":370432,"commit":"a868dbdde8d02690210420f315457c35e3aa3331"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/370184","number":370184,"commit":"08a46335cd3a4e04c0c2203349b271dbee9ba9bd"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369889","number":369889,"commit":"d7eb4e824870baf5010d2d432dec3fd05bb26a05"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369876","number":369876,"commit":"02ff815e179f596431a507ecc80b22fa1c648b32"}],"number":370624}
buildkite-pr-comment-->

h88 committed:
jest --fix

h88 committed:
oneChat tool test alignment with new wf inputs mech to handle JSON Schema array/object inputs

h88 committed:
unwrapping refine effects in JSON Schema generation for Monaco YAML autocomplete

h88 commented on src/platform/packages/shared/kbn-workflows/spec/schema.ts:
```diff
@@ -422,7 +422,55 @@ export const WorkflowInputSchema = z.union([
+// structured data models.
+export const JsonModelSchema = z
+  .object({
+    properties: z.record(z.string(), z.any()).optional(),
```
8318d8b74f92ced18f0f658581ae9e43f191b4a0

I agree ±

In the referenced commit:
1. Unwraped `refine` effects in `get_workflow_json_schema.ts` to expose the base schema for JSON Schema generation
2. Added `type` to `JsonModelSchema` root for Monaco clarity

This minimal change ensures:
-  Base schema structure is properly exposed to Monaco
-  Monaco receives `type: 'object'` for better autocomplete
-  Validation via `refine` still works correctly
-  No breaking changes

## Why Not Full Recursive Definition (Yet)

A full recursive Zod definition for JSON Schema Draft 7 would require 1000+ lines to model all types, combinators, and edge cases. The current approach with `refine` + unwrapping provides:
- Minimal changes and side effects
- Handles all JSON Schema edge cases via Ajv validation
- Flexible - accepts any valid JSON Schema

We can migrate to a recursive definition later if Monaco autocomplete for nested schemas becomes critical, or track in a separate PR? but this commit addresses the immediate issue with minimal risk.

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

# Conflicts:
#	src/platform/packages/shared/kbn-workflows/spec/lib/get_workflow_json_schema.kibana.test.ts

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/372812)
* Commit: 7e361667878ef462695d1050c0f8ead5ade95dc7

### Failed CI Steps
* [FTR Configs #41](https://buildkite.com/elastic/kibana-pull-request/builds/372812#019b1c94-6bdd-4511-abfa-76fb6b9b972d)
* [Agent Builder API Smoke Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/372812#019b1c73-3c1b-4e36-bb28-0c8f42a39985)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/372812#019b1c73-3c1b-4e36-bb28-0c8f42a39985) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/372812/jobs/019b1c73-3c1b-4e36-bb28-0c8f42a39985/artifacts/019b1c96-4ebe-432a-8903-0b52535ebb12) Agent Builder API Smoke Tests #<span></span>1 / Agent Builder - LLM Smoke tests Connector "bedrock-claude-3-5-sonnet" Converse API sync can execute a tool
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/372812#019b1c94-6bdd-4511-abfa-76fb6b9b972d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/372812/jobs/019b1c94-6bdd-4511-abfa-76fb6b9b972d/artifacts/019b1cbf-d03c-4912-a7df-b560f3b340f1) FTR Configs #<span></span>41 / Options list control Allow expensive queries setting is off Can search options list for available options - does not find partial match

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/fe9a41615e3f2bf7442ceae3d50325789751f6f1) | [after](https://github.com/elastic/kibana/commit/7e361667878ef462695d1050c0f8ead5ade95dc7) | diff |
| --- | --- | --- | --- |
| `workflowsManagement` | 1114 | 1233 | +119 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/fe9a41615e3f2bf7442ceae3d50325789751f6f1) | [after](https://github.com/elastic/kibana/commit/7e361667878ef462695d1050c0f8ead5ade95dc7) | diff |
| --- | --- | --- | --- |
| `@kbn/workflows` | 363 | 360 | -3 |

#### Any counts in public APIs
> Total count of every any typed public API. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats any` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/fe9a41615e3f2bf7442ceae3d50325789751f6f1) | [after](https://github.com/elastic/kibana/commit/7e361667878ef462695d1050c0f8ead5ade95dc7) | diff |
| --- | --- | --- | --- |
| `@kbn/workflows` | 72 | 75 | +3 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/fe9a41615e3f2bf7442ceae3d50325789751f6f1) | [after](https://github.com/elastic/kibana/commit/7e361667878ef462695d1050c0f8ead5ade95dc7) | diff |
| --- | --- | --- | --- |
| `onechat` | 481.9KB | 482.7KB | +894.0B |
| `workflowsManagement` | 1.2MB | 1.5MB | :warning: +290.6KB |
| total |  |  | +291.5KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/fe9a41615e3f2bf7442ceae3d50325789751f6f1) | [after](https://github.com/elastic/kibana/commit/7e361667878ef462695d1050c0f8ead5ade95dc7) | diff |
| --- | --- | --- | --- |
| `workflowsManagement` | 7.7KB | 7.7KB | -1.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/fe9a41615e3f2bf7442ceae3d50325789751f6f1) | [after](https://github.com/elastic/kibana/commit/7e361667878ef462695d1050c0f8ead5ade95dc7) | diff |
| --- | --- | --- | --- |
| `@kbn/workflows` | 397 | 394 | -3 |

#### ESLint disabled in files
| id | [before](https://github.com/elastic/kibana/commit/fe9a41615e3f2bf7442ceae3d50325789751f6f1) | [after](https://github.com/elastic/kibana/commit/7e361667878ef462695d1050c0f8ead5ade95dc7) | diff |
| --- | --- | --- | --- |
| `workflowsManagement` | 37 | 36 | -1 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/fe9a41615e3f2bf7442ceae3d50325789751f6f1) | [after](https://github.com/elastic/kibana/commit/7e361667878ef462695d1050c0f8ead5ade95dc7) | diff |
| --- | --- | --- | --- |
| `@kbn/workflows` | 19 | 26 | +7 |
| `workflowsManagement` | 86 | 93 | +7 |
| total |  |  | +14 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/fe9a41615e3f2bf7442ceae3d50325789751f6f1) | [after](https://github.com/elastic/kibana/commit/7e361667878ef462695d1050c0f8ead5ade95dc7) | diff |
| --- | --- | --- | --- |
| `@kbn/workflows` | 25 | 32 | +7 |
| `workflowsManagement` | 123 | 129 | +6 |
| total |  |  | +13 |


</details>


### History
* :green_heart: [Build #371684](https://buildkite.com/elastic/kibana-pull-request/builds/371684) succeeded 8318d8b74f92ced18f0f658581ae9e43f191b4a0
* :green_heart: [Build #370737](https://buildkite.com/elastic/kibana-pull-request/builds/370737) succeeded e1387d041e66831c260bec607adca1e4ac72cef8
* :broken_heart: [Build #370624](https://buildkite.com/elastic/kibana-pull-request/builds/370624) failed a40e23766e12f69c1ec69a0eee585d76132a9e6e
* :broken_heart: [Build #370587](https://buildkite.com/elastic/kibana-pull-request/builds/370587) failed f5e4820dee47a1b5bfb8282cb37fcbd7ebefc292
* :yellow_heart: [Build #370432](https://buildkite.com/elastic/kibana-pull-request/builds/370432) was flaky a868dbdde8d02690210420f315457c35e3aa3331
* :broken_heart: [Build #370184](https://buildkite.com/elastic/kibana-pull-request/builds/370184) failed 08a46335cd3a4e04c0c2203349b271dbee9ba9bd

cc @h88

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372812","number":372812,"commit":"7e361667878ef462695d1050c0f8ead5ade95dc7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371684","number":371684,"commit":"8318d8b74f92ced18f0f658581ae9e43f191b4a0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/370737","number":370737,"commit":"e1387d041e66831c260bec607adca1e4ac72cef8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/370624","number":370624,"commit":"a40e23766e12f69c1ec69a0eee585d76132a9e6e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/370587","number":370587,"commit":"f5e4820dee47a1b5bfb8282cb37fcbd7ebefc292"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/370432","number":370432,"commit":"a868dbdde8d02690210420f315457c35e3aa3331"}],"number":372812}
buildkite-pr-comment-->

skynetigor commented on src/platform/packages/shared/kbn-workflows/spec/schema.ts:
```diff
@@ -422,7 +422,56 @@ export const WorkflowInputSchema = z.union([
+// structured data models.
+export const JsonModelSchema = z
+  .object({
+    type: z.literal('object').optional(),
```
Then it means I can't use JsonModelSchema as schema for outputSchema in ai.prompt step, because that will limit the user to using only object in the root of schema.
What would be a good solution to overcome it so that the user can also use array schema in the root of ai.prompt outputSchema?

This is what I need to let the user to write:
```yaml
  - name: analysis
    type: inference.completion
    connector-id: 514e7afd-5461-4398-b0b6-80dccf09c466
    with:
      outputSchema:
         title: Person Array
         type: array
         items:
           title: Person
           type: object
           properties:
             name:
               type: string
               description: The person's first name.
             surname:
               type: string
               description: The person's last name.


      input: |
        - Output JSON array with multiple object
        - Generate random count of elements from 3 to 10
        - Generate some famous name and surname (Brad Pitt, etc)
        - Don't include anything else except JSON into the response

```

skynetigor reviewed (approved):
LGTM.
There is only one concern regarding root JSON schema type that's currently hardcoded to "object".

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

# Conflicts:
#	src/platform/packages/shared/kbn-workflows/spec/lib/generate_yaml_schema_from_connectors.ts
#	src/platform/packages/shared/kbn-workflows/spec/schema.test.ts

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

# Conflicts:
#	x-pack/platform/plugins/shared/agent_builder/public/application/components/tools/execute/test_tools.tsx

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

# Conflicts:
#	src/platform/plugins/shared/workflows_execution_engine/server/workflow_context_manager/build_workflow_context.ts
#	src/platform/plugins/shared/workflows_execution_engine/server/workflow_context_manager/tests/workflow_context_manager.test.ts

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

h88 committed:
minor fixes

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

# Conflicts:
#	src/platform/packages/shared/kbn-workflows/spec/lib/generate_yaml_schema_from_connectors.ts
#	src/platform/plugins/shared/workflows_management/public/features/validate_workflow_yaml/lib/use_yaml_validation.ts
#	src/platform/plugins/shared/workflows_management/public/features/validate_workflow_yaml/model/types.ts
#	src/platform/plugins/shared/workflows_management/public/widgets/workflow_yaml_editor/lib/autocomplete/get_completion_item_provider.ts
#	src/platform/plugins/shared/workflows_management/public/widgets/workflow_yaml_editor/lib/autocomplete/suggestions/get_suggestions.ts

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/383483)
* Commit: 9b9185abda9e97ed5592850945847fd733a65d37

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/383483#019bd81e-2cdc-4f04-8354-1fbe9d962897)

### History
* :broken_heart: [Build #383469](https://buildkite.com/elastic/kibana-pull-request/builds/383469) failed bcbd72743f5038badcf4a5531575ce70e9cd1fef
* :yellow_heart: [Build #382881](https://buildkite.com/elastic/kibana-pull-request/builds/382881) was flaky e2fc64510555bdf2f4eb8781d6a3f977a61beea3
* :yellow_heart: [Build #382268](https://buildkite.com/elastic/kibana-pull-request/builds/382268) was flaky 5028d91c49488945ae7171d973c3fc5891dacf63
* :yellow_heart: [Build #380810](https://buildkite.com/elastic/kibana-pull-request/builds/380810) was flaky 4d2ca1ea881201f7f265611eba43162307e88e5b
* :broken_heart: [Build #380458](https://buildkite.com/elastic/kibana-pull-request/builds/380458) failed e88a4efa73de46327f813192d52dd79a18e79b5b
* :green_heart: [Build #378248](https://buildkite.com/elastic/kibana-pull-request/builds/378248) succeeded f1166b15bd26f46760ada0668e6d164d27e19f42

cc @h88

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/383483","number":383483,"commit":"9b9185abda9e97ed5592850945847fd733a65d37"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/383469","number":383469,"commit":"bcbd72743f5038badcf4a5531575ce70e9cd1fef"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/382881","number":382881,"commit":"e2fc64510555bdf2f4eb8781d6a3f977a61beea3"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/382268","number":382268,"commit":"5028d91c49488945ae7171d973c3fc5891dacf63"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/380810","number":380810,"commit":"4d2ca1ea881201f7f265611eba43162307e88e5b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/380458","number":380458,"commit":"e88a4efa73de46327f813192d52dd79a18e79b5b"}],"number":383483}
buildkite-pr-comment-->

h88 committed:
Merge branch 'main' into 1wf-json-schema-inputs

h88 committed:
Integrate json-schema-inputs with Zod/4.2/fromJSONSchema polyfill
